core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java

```
@@ -0,0 +1,192 @@
+import org.elasticsearch.index.mapper.ParseContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Utility to extract query terms from queries and create queries from documents.
+ */
+public class QueryMetadataService {
+
+    public static String QUERY_METADATA = "_query_metadata_";
+    public static String QUERY_METADATA_FIELD_PREFIX = QUERY_METADATA + "field_";
+    public static String QUERY_METADATA_FIELD_UNKNOWN = QUERY_METADATA + "unknown";
+    public static FieldType QUERY_METADATA_FIELD_TYPE = new FieldType();
+
+    static {
+        QUERY_METADATA_FIELD_TYPE.setTokenized(false);
+        QUERY_METADATA_FIELD_TYPE.setIndexOptions(IndexOptions.DOCS);
+        QUERY_METADATA_FIELD_TYPE.freeze();
+    }
+
+    /**
+     * Extracts all terms from the specified query and adds it to the specified document.
+     */
+    public void extractQueryMetadata(Query query, ParseContext.Document document) {
+        List<Term> queryTerms;
+        try {
+            queryTerms = extractQueryMetadata(query);
+        } catch (IllegalArgumentException e) {
+            document.add(new Field(QUERY_METADATA_FIELD_UNKNOWN, new BytesRef(), QUERY_METADATA_FIELD_TYPE));
+            return;
+        }
+        for (Term term : queryTerms) {
+            document.add(new Field(QUERY_METADATA_FIELD_PREFIX + term.field(), term.bytes(), QUERY_METADATA_FIELD_TYPE));
+        }
+    }
+
+    /**
+     * Extracts all query terms from the provided query and adds it to specified list.
+     *
+     * From boolean query with no should clauses or phrase queries only the the longest term are selected,
+     * since that those terms are likely to be the rarest. Boolean query's must_not clauses are always ignored ignored.
+     *
+     * If from part of the query, no query terms can be extracted then term extraction is stopped and
+     * an IllegalArgumentException is thrown.
```

maybe we need a more specific exception? I'm afraid that if eg. one of the methods that extractQueryMetadata calls throws an IAE then term extraction will silently fail